### PR TITLE
Feature/hierarchical graphs

### PR DESCRIPTION
**Pull request by icedoom888**
An implementation of hierarchical graph model.
This allows for the creation of a model capable of dealing with multi-layered hierarchical graphs.
The model is agnostic to the choosen architecture (GNN, GraphTransformer).

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-models?pullRequest=37) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-models?pullRequest=37) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-models?pullRequest=37) it.</sub>

**Comment by mchantry** [comment]
@icedoom888 thanks for the contribution.
Could you rebase it on top of
https://github.com/ecmwf/anemoi-models/pull/45/files which should be merged in the next day.
to solve the instantiate issue?

File: src/anemoi/models/layers/mapper.py [hunk 1]
```diff
@@ -134,8 +135,8 @@ def _register_edges(
         trainable_size : int
             Trainable tensor size
         """
-        assert sub_graph, f"{self.__class__.__name__} needs a valid sub_graph to register edges."
```
Comment (floriankrb) [inline src/anemoi/models/layers/mapper.py:137]: why removing the check on `sub_graph`?

**Comment by icedoom888** [comment]
Hello @mchantry, @floriankrb, I am currently blocked due to the switch to anemoi-training from aifs-mono. 
Will resume working on this and address your comments once my pipeline is functional again!

**Comment by icedoom888** [comment]
UPDATE:
- Full refactor of hierarchical model implementation  (done w/ @JPXKQX )
- More readable code 
- Moving to anemoi-training for model testing

Please feel free to take a look and make any sort of suggestions!

File: src/anemoi/models/models/hierarchical.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
+# (C) Copyright 2024 ECMWF.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+#
+
```
Comment (HCookie) [inline src/anemoi/models/models/hierarchical.py:9]: Can the copyright statement be updated in accordance with https://github.com/ecmwf/anemoi-models/pull/67?

File: src/anemoi/models/models/hierarchical.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        ----------
+        config : DotDict
+            Job configuration
+        data_indices : dict
+            Data indices
+        graph_data : HeteroData
+            Graph definition
+        """
+        nn.Module.__init__(self)
+
```
Comment (HCookie) [inline src/anemoi/models/models/hierarchical.py:51]: Why does this subclass of `AnemoiModelEncProcDec`, not call its `super().__init__` method?
There appears to be a chunk of copied code from the parent, particularly the `boundings`.

If there are constraints on how the parent inits, then a refactor for an abstract base class would make sense.

File: src/anemoi/models/models/hierarchical.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        for hidden in self._graph_hidden_names:
+            self._register_latlon(hidden, hidden)
+
+        input_dim = self.multi_step * self.num_input_channels + self.latlons_data.shape[1] + self.trainable_data_size
+
+        print(
+            input_dim,
+            getattr(self, f"latlons_{self._graph_hidden_names[0]}").shape[1],
+            self.hidden_dims[self._graph_hidden_names[0]],
+        )
```
Comment (HCookie) [inline src/anemoi/models/models/hierarchical.py:86]: Is this a leftover debug?

File: src/anemoi/models/models/hierarchical.py (hunk 3 above)
Comment (JPXKQX) [reply src/anemoi/models/models/hierarchical.py:52]: This will be addressed in PR #79

File: src/anemoi/models/models/hierarchical.py (hunk 2 above)
Comment (icedoom888) [reply src/anemoi/models/models/hierarchical.py:9]: Done ✔️

File: src/anemoi/models/models/hierarchical.py (hunk 4 above)
Comment (icedoom888) [reply src/anemoi/models/models/hierarchical.py]: Removed ✔️

File: src/anemoi/models/layers/processor.py [hunk 5]
```diff
@@ -323,6 +323,7 @@ def forward(
         *args,
         **kwargs,
     ) -> Tensor:
+
```
Comment (JPXKQX) [inline src/anemoi/models/layers/processor.py:326]:
```suggestion
```

**Review by JPXKQX — APPROVED** [review]

File: src/anemoi/models/models/hierarchical.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        self._assert_matching_indices(data_indices)
+        self.data_indices = data_indices
+
+        self.multi_step = model_config.training.multistep_input
+
+        # self.node_attributes = {hidden_name: NamedNodesAttributes(model_config.model.trainable_parameters[hidden_name], self._graph_data)
+        #                         for hidden_name in self._graph_hidden_names}
+        self.node_attributes = NamedNodesAttributes(model_config.model.trainable_parameters.hidden, self._graph_data)
+
+        input_dim = self.multi_step * self.num_input_channels + self.latlons_data.shape[1] + self.trainable_data_size
```
Comment (JPXKQX) [inline src/anemoi/models/models/hierarchical.py:76]:
```suggestion
        input_dim = self.multi_step * self.num_input_channels + self.node_attributes.attr_ndims[self._graph_name_data]
```

File: src/anemoi/models/models/hierarchical.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+        # Instantiation of model output bounding functions (e.g., to ensure outputs like TP are positive definite)
+        self.boundings = nn.ModuleList(
+            [
+                instantiate(cfg, name_to_index=self.data_indices.internal_model.output.name_to_index)
+                for cfg in getattr(model_config.model, "bounding", [])
+            ]
+        )
+
+    def _create_trainable_attributes(self) -> None:
```
Comment (JPXKQX) [inline src/anemoi/models/models/hierarchical.py:173]: The function `create_trainable_attributes()` can be removed. Functionality moved to self.nodes_attributes

File: src/anemoi/models/models/hierarchical.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+    def forward(self, x: Tensor, model_comm_group: Optional[ProcessGroup] = None) -> Tensor:
+        batch_size = x.shape[0]
+        ensemble_size = x.shape[2]
+
+        # add data positional info (lat/lon)
+        x_trainable_data = torch.cat(
+            (
+                einops.rearrange(x, "batch time ensemble grid vars -> (batch ensemble grid) (time vars)"),
+                self.trainable_data(self.latlons_data, batch_size=batch_size),
```
Comment (JPXKQX) [inline src/anemoi/models/models/hierarchical.py:191]:
```suggestion
                self.node_attributes(self._graph_name_data, batch_size=batch_size),
```

**Review by JPXKQX — CHANGES_REQUESTED** [review]

**Review by JPXKQX — APPROVED** [review]

**Pull request merged**